### DISCUSSION TITLE
HugSQL support

### DISCUSSION THREAD

**Discussion by borkdude** [babashka/babashka#945] [Ideas]
@arichiardi We were discussing HugSQL support in https://github.com/borkdude/sci/pull/547

Let's continue the discussion here.

Adding a few functions from tools.reader makes bb compatible with hugsql.core. Tools.reader is already well within the transitive dependency graph of babashka so exposing part of it doesn't add a lot of binary size.

I exposed just enough to make it compatible with hugsql.core and got this working:

``` 
/tmp/hugsql (master) $ bb -cp hugsql-core/src:hugsql-adapter/src -e "(require '[hugsql.core :as h])" -e '(def m (h/map-of-sqlvec-fns-from-string "-- :name hello\nselect * from foo where x = :x")) (-> m :hello-sqlvec :fn (apply [{:x 1}]))'
["select * from foo where x = ?" 1]
``` 

Linux: https://21179-201467090-gh.circle-artifacts.com/0/release/babashka-0.5.1-SNAPSHOT-linux-amd64-static.tar.gz

MacOS: https://21183-201467090-gh.circle-artifacts.com/0/release/babashka-0.5.1-SNAPSHOT-macos-amd64.tar.gz

**Comment by arichiardi** [comment]
Oh this is cool, I will try that out for sure!

**Comment by liny01-nbsa** [comment]
Hi @borkdude , this looks great. I tried to follow as you have shown, please guide me if I'm mistaken:

```
➜  tmp git clone https://github.com/layerware/hugsql.git
Cloning into 'hugsql'...
remote: Enumerating objects: 2022, done.
remote: Counting objects: 100% (187/187), done.
remote: Compressing objects: 100% (140/140), done.
remote: Total 2022 (delta 42), reused 155 (delta 29), pack-reused 1835
Receiving objects: 100% (2022/2022), 1.45 MiB | 6.51 MiB/s, done.
Resolving deltas: 100% (1090/1090), done.
➜  tmp cd hugsql
➜  hugsql git:(master) bb --version
babashka v0.7.3
➜  hugsql git:(master) bb -cp hugsql-core/src:hugsql-adapter/src -e "(require '[hugsql.core :as h])" -e '(def m (h/map-of-sqlvec-fns-from-string "-- :name hello\nselect * from foo where x = :x")) (-> m :hello-sqlvec :fn (apply [{:x 1}]))'
----- Error --------------------------------------------------------------------
Type:     java.lang.Exception
Message:  Could not find namespace: clojure.tools.reader.reader-types.
Location: xxx/tmp/hugsql/hugsql-core/src/hugsql/parser.clj:2:3

----- Context ------------------------------------------------------------------
1: (ns hugsql.parser
2:   (:require [clojure.string :as string]
     ^--- Could not find namespace: clojure.tools.reader.reader-types.
3:             [clojure.tools.reader.reader-types :as r]))
4:
5: (defn- parse-error
6:   ([rdr msg]
7:    (parse-error rdr msg {}))

----- Stack trace --------------------------------------------------------------
hugsql.parser - xxx/tmp/hugsql/hugsql-core/src/hugsql/parser.clj:2:3
hugsql.core   - xxx/tmp/hugsql/hugsql-core/src/hugsql/core.clj:2:3
```

I also tested against a commit of hugsql version that's before Jul 16, 2021, as last message was. However, I'm getting the same error.

Is there a test version of bb I can use to see how it works? The links you provide no long exist.

Thanks!

**Reply by borkdude** [reply]
@liny01-nbsa What operating system are you on? Let me try to get the branch up to date.

**Reply by borkdude** [reply]
Anyway, here is an updated macOS link: https://25689-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.4-SNAPSHOT-macos-amd64.tar.gz

And linux: https://25688-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.4-SNAPSHOT-linux-amd64-static.tar.gz

**Reply by liny01-nbsa** [reply]
@borkdude , I work on Linux and WSL 2. thanks for the package, I have tested the case you provided, it's working perfectly.

I will start testing with some existing HugSQL files we have, will post back anything I find.

**Reply by liny01-nbsa** [reply]
Hi @borkdude, good news, most of my tests are successful, there're a few issues, some of them could be the SQL pods related, I'm working to see if I can resolve them all. We are very happy that one of our most complex query, a ~700 line, CTE nested one, processed properly. I will post back when we finish testing all our remaining SQL.

**Reply by liny01-nbsa** [reply]
@borkdude, I have finished my testing. Everything looks good for our use cases.

**Reply by borkdude** [reply]
@liny01-nbsa Thanks for testing, I merged the hugsql branch to master so it will be available in the next release.

Meanwhile you can use https://github.com/babashka/babashka-dev-builds 

As soon as the current master build is done, you can download the binaries from there.

**Reply by liny01-nbsa** [reply]
Thanks, @borkdude, confirmed build `0.7.8-SNAPSHOT` works as expected.
